mesonbuild/modules/unstable_wayland.py

```diff
@@ -0,0 +1,132 @@
+        for arg in args:
+            if isinstance(arg, str):
+                if os.path.isabs(arg):
+                    xml_path = arg
+                else:
+                    xml_path = os.path.join(self.interpreter.environment.source_dir, arg)
+                    if not os.path.exists(xml_path):
+                        raise MesonException(f'The file {xml_path} does not exist.')
+            else:
+                xml_path = arg.absolute_path(
+                    self.interpreter.environment.source_dir,
+                    self.interpreter.environment.build_dir)
+
+            xml_file = os.path.basename(xml_path)
+            name, _ = os.path.splitext(xml_file)
+
+            code = CustomTarget(
+                f'{name}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{scope}-code', '@INPUT@', '@OUTPUT@'],
+                [xml_path],
+                [f'{name}-protocol.c'],
+                backend=state.backend,
+            )
+            targets.append(code)
+
+            header = CustomTarget(
+                f'{name}-{side}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{side}-header', '@INPUT@', '@OUTPUT@'],
+                [xml_path],
+                [f'{name}-{side}-protocol.h'],
+                backend=state.backend,
+            )
+            targets.append(header)
+
+        return ModuleReturnValue(targets, [targets])
+
+    def find_protocol(self, state, args, kwargs):
+        if len(args) != 1 or not isinstance(args[0], str):
+            raise MesonException('find requires exactly one string argument: the protocol name.')
+
+        base_name = args[0]
+        xml_state = kwargs.get('state', 'stable')
+
+        if (xml_state == 'staging' or xml_state == 'unstable') and 'version' not in kwargs:
+            raise MesonException(f'{xml_state} protocols require a version number.')
+
+        if self.protocols_dep is None:
+            self.protocols_dep = self.interpreter.func_dependency(state.current_node, ['wayland-protocols'], {})
+
+        if self.pkgdatadir is None:
+            self.pkgdatadir = self.protocols_dep.get_variable(pkgconfig='pkgdatadir', internal='pkgdatadir')
+
+        if xml_state == 'stable':
+            xml_name = f'{base_name}.xml'
+        elif xml_state == 'staging':
+            version = kwargs['version']
+            xml_name = f'{base_name}-v{version}.xml'
+        elif xml_state == 'unstable':
+            version = kwargs['version']
+            xml_name = f'{base_name}-unstable-v{version}.xml'
+        else:
+            raise MesonException('stage must be either stable, staging, or unstable.')
+
+        path = os.path.join(self.pkgdatadir, xml_state, base_name, xml_name)
+
+        if not os.path.exists(path):
+            raise MesonException(f'The file {path} does not exist.')
+
+        return ModuleReturnValue(path, [path])
```

Comment:
I think you should return `File.from_absolute_file(path)`, better return a file object than a string. Also ModuleReturnValue is only needed when returning build targets.